### DISCUSSION TITLE
Natural Sort Columns

### DISCUSSION THREAD

**Discussion by intrepidws** [filamentphp/filament#1835] [Help] [answered]
I have a resource with a field that contains a list of items that have both numbers and letters. When sorting this column, the order isn't correct. Example:

1. CAST 1200
2. CAST 300
3. CAST 600
4. CAST 900

This is a common MySQL problem but a workaround fix [does exist](https://www.copterlabs.com/natural-sorting-in-mysql/). So in my case this would result in the following SQL query:

`SELECT name FROM transformers ORDER by LENGTH(name), name`

Or for DESC:

`SELECT name FROM transformers ORDER by LENGTH(name) DESC, name DESC`

Thoughts on allowing this type of Natural Sort?

**Comment by danharrin** [comment]
Is this just an issue with MySQL? Is there a consistent way to fix it for all Laravel DB drivers?

**Reply by intrepidws** [reply]
There are ways in Laravel to sort collections and arrays using a natural sort flag, but not in an Eloquent query. I don't know whether that might be something worth PR'ing to Laravel itself but I don't think that's something I'm capable of doing myself. I'm also not sure how universal of a "fix" the above method actually is.

**Comment by intrepidws** [comment]
As it turns out, you can pass a `DB::raw()` statement to the `sortable()` method. I was able to solve my issue using the following:

```php
Tables\Columns\TextColumn::make('name')
    ->searchable()
    ->sortable([\DB::raw('LENGTH(name)'), 'name']),
```

**Reply by intrepidws** [reply]
Well, that was a big failure. You *can* use the `DB::raw()` statement, which is actually great to know and very helpful, BUT the solution (ordering by `LENGTH(name)` in MySQL) doesn't actually achieve the desired result for my data set.

**Reply by ryangjchandler** [reply]
I guess it's because the lengths aren't actually different between `600` and `300` right? If you had something like `12` and then `136`, it'd be order correctly (I think).

**Reply by intrepidws** [reply]
It's more complicated than that unfortunately, and it's giving me a headache haha. Apparently this is the best solution... and it's kinda crazy. I actually can't even get it to work for some reason.

https://stackoverflow.com/questions/153633/natural-sort-in-mysql/12257917#12257917

**Reply by danharrin** [reply]
That's disgusting lol, especially just for a simple order clause.

**Reply by intrepidws** [reply]
It looks like Filament 3 no longer supports passing a `DB::raw()` statement to the `sortable()` method? I'm upgrading this codebase to Filament 3 and getting the following error:

```php
Filament\Tables\Columns\Column::getSortColumnForQuery(): Argument #2 ($sortColumn) must be of type string, Illuminate\Database\Query\Expression given, called in /Users/iwsmacstudio/Code/cw-landscape-lighting-wattson/vendor/filament/tables/src/Columns/Concerns/InteractsWithTableQuery.php on line 138
```

I've tried modifying my code like this:

```php
Tables\Columns\TextColumn::make('name')
    ->searchable()
    ->sortable([\DB::raw('LENGTH(name)')->getValue(new Grammar()), 'name']),
```

The resulting SQL has the `LENGTH(name)` portion within quotes, so the database looks for a field with that name, rather than running the function.

I've started looking through the InteractsWithTableQuery.php file to look for what may have changed between 2.x and 3.x but so far having trouble finding anything.

Any thoughts?

**Reply by intrepidws** [reply]
I've dug a bit deeper here and if you change the following method declaration:

```php
protected function getSortColumnForQuery(EloquentBuilder $query, string $sortColumn, ?array $relationships = null): string | Builder
```

To this, then it works.

```php
use Illuminate\Database\Query\Expression as ExpressionQuery;

protected function getSortColumnForQuery(EloquentBuilder $query, string | ExpressionQuery $sortColumn, ?array $relationships = null): string | Builder | ExpressionQuery
```

I'm happy to contribute the PR for this, but I'm concerned that maybe I'm missing something about why this was changed from 2.x to 3.x. Let me know what you think.

**Reply by danharrin** [reply]
I think you should just pass a function to `sortable()` to modify the `$query`

**Reply by intrepidws** [reply]
I tried that, this is the code:

```php
Tables\Columns\TextColumn::make('name')
    ->sortable(query: function (Builder $query, string $direction): Builder {
        return $query->orderByRaw(\DB::raw('LENGTH(name)'));
    }),
```

That returns a similar error:

```php
Object of class Illuminate\Database\Query\Expression could not be converted to string
```

**Reply by intrepidws** [reply]
Wait, my bad - I doubled up the "raw" portion. This works:

```php
Tables\Columns\TextColumn::make('name')
    ->sortable(query: function (Builder $query, string $direction): Builder {
        return $query->orderByRaw('LENGTH(name)');
    }),
```

**Comment by intrepidws** [comment, marked as answer]
OK so for some closure here. Unless and until Laravel implements a solution, which seems very unlikely, then the solution is a bit convoluted but does work. Basically you need to create a couple of stored MySQL functions, which are [outlined here](https://stackoverflow.com/questions/153633/natural-sort-in-mysql/12257917#12257917). I had to create these manually via my DB GUI and couldn't get it to work within a migration for some reason.

With those functions created, my final Filament 3.x code was:

```php
Tables\Columns\TextColumn::make('name')
    ->sortable(query: function (Builder $query, string $direction): Builder {
        return $query->orderByRaw('udf_NaturalSortFormat(name, 10, ".")');
    }),
```

Big shout out to @danharrin for his help in upgrading my old 2.x version of this code to work in 3.x.

For anyone who needs it, my final Filament 2.x code is below. It's possible that the 3.x code above will work as well, but I haven't tested it with 2.x.

```php
Tables\Columns\TextColumn::make('name')
    ->searchable()
    ->sortable([\DB::raw('udf_NaturalSortFormat(name, 10, ".")')]),
```

**Comment by aksalsf** [comment]
I am using this way for my case. Hopefully can be helpful,

```
TextColumn::make('product_id')
                ->label('Product')
                ->sortable(['product_id', Application::raw('LENGTH(product_id)')]),
```

This way print this query

```
... order by LENGTH(product_id) asc/desc, `product_id` asc/desc
```

> **Note**
> asc/desc in query generated automatically by the sort action


My dataset structure:

```
1-PRODUCT-2022
2-PRODUCT-2022
...
16-PRODUCT-2022
```
